azext_iot/operations/hub.py

```diff
@@ -1082,15 +1089,22 @@ def _iot_key_regenerate_batch(
         else:
             batches.append(items[:])
             items = []
-    for i in tqdm(range(len(batches)), desc="Bulk key regeneration is in progress", ascii=' #'):
+    for batch in tqdm(batches, desc="Bulk key regeneration is in progress", ascii=' #'):
         # call
-        try:
-            result = service_sdk.service.bulk_regenerate_device_key_method(
-                policy_key=renew_key_type,
-                devices=batches[i]
-            )
-        except CloudError as e:
-            handle_service_exception(e)
+        result = None
+        tries = 0
+        while tries < 3:
+            try:
+                result = service_sdk.service.bulk_regenerate_device_key_method(
+                    policy_key=renew_key_type,
+                    devices=batch
+                )
+                break
+            except CloudError as e:
+                tries += 1
+                if tries == 3 or e.status_code != 429:
+                    handle_service_exception(e)
+                sleep(20)
```

Comment:
I think we should use constants for the retry values and the sleep value, or at least define them in one place so they're easier to tweak if needed.

Also, I'm not super familiar with the rate of calls in these key updates vs the API limits - 20s seems like a high backoff?

Finally - if we fail out of this through too many retries, we wouldn't have any way to know if any devices had already been rotated... Do you think it's of any value to log the succeeded batches?

Reply:
agree on constants tho it is a bit awk with this file

for the 20 seconds - the error does mention 10 seconds but 20 works better (as in I have gotten to the end of the retries with 10 second sleeps + 15 second sleeps)

I will add a logger.warning with the rotated keys